engine/consensus/dkg/reactor_engine.go

```diff
@@ -229,42 +229,59 @@ func (e *ReactorEngine) startDKGForEpoch(currentEpochCounter uint64, first *flow
 	e.registerPhaseTransition(curDKGInfo.phase3FinalView, dkgmodule.Phase3, e.end(nextEpochCounter))
 }
 
-// handleEpochCommittedPhaseStarted checks that the local DKG completed and
-// that our locally computed key share is consistent with the canonical key vector.
+// handleEpochCommittedPhaseStarted is invoked upon the transition to the EpochCommitted
+// phase, when the canonical beacon key vector is incorporated into the protocol state.
 //
-// If our key is inconsistent, we log a warning message.
-// CAUTION: we will still sign using the inconsistent key, to be fixed with Crypto
-// v2 by falling back to staking signatures in this case.
+// This function checks that the local DKG completed and that our locally computed
+// key share is consistent with the canonical key vector. When this function returns,
+// an end state for the just-completed DKG is guaranteed to be stored (if not, the
+// program will crash). Since this function is invoked synchronously before the end
+// of the current epoch, this guarantees that when we reach the end of the current epoch
+// we will either have a usable beacon key (successful DKG) or a DKG failure end state
+// stored, so we can safely fall back to using our staking key.
 //
-func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64) {
+// CAUTION: This function is not safe for concurrent use. This is not enforced within
+// the ReactorEngine - instead we rely on the protocol event emission being single-threaded
+//
+func (e *ReactorEngine) handleEpochCommittedPhaseStarted(currentEpochCounter uint64, firstBlock *flow.Header) {
 
-	// the epoch counter the DKG just completed is for
-	dkgEpochCounter := currentEpochCounter + 1
+	// the DKG we have just completed produces keys that we will use in the next epoch
+	nextEpochCounter := currentEpochCounter + 1
 
 	log := e.log.With().
 		Uint64("cur_epoch", currentEpochCounter). // the epoch we are in the middle of
-		Uint64("dkg_for_epoch", dkgEpochCounter). // the epoch the just-finished DKG was preparing for
+		Uint64("next_epoch", nextEpochCounter).   // the epoch the just-finished DKG was preparing for
 		Logger()
 
-	// TODO - first check whether we've already stored the dkg end state, since
-	// we need to handle multiple calls to this method
+	// Check whether we have already set the end state for this DKG.
+	// This can happen if the DKG failed locally, if we failed to generate
+	// a local private beacon key, or if we crashed while performing this
+	// check previously.
+	endState, err := e.dkgState.GetDKGEndState(nextEpochCounter)
+	if err == nil {
+		log.Warn().Msgf("already set dkg end state when checking beacon key consistency: %s", endState.String())
```

Comment:
I think the end state could be set by `(e *ReactorEngine) end` to `DKGEndStateDKGFailure`. I mean the state was not necessarily set by a previous consistency check right? 
```suggestion
		log.Warn().Msgf("already set dkg end state: %s", endState.String())
```

Reply:
https://github.com/onflow/flow-go/pull/1732/commits/44d144ccff93dd705c1e0e075f8d3375a047accc